### PR TITLE
Added copy and move constructors in Reparametrization layer

### PR DESCRIPTION
**Pull request by Anmol2001**
This PR is regarding the issue #2625 and I have added copy and move constructors for `Reparametrization`  layer and also added the corresponding tests.

### CODE REVIEW

File: src/mlpack/methods/ann/layer/reparametrization_impl.hpp [hunk 1]
```diff
@@ -46,7 +46,60 @@ Reparametrization<InputDataType, OutputDataType>::Reparametrization(
...
+         includeKl = layer.includeKl;
+         beta = layer.beta;
+       }
+       return *this;
+}
+    
+template <typename InputDataType, typename OutputDataType>
+Reparametrization<InputDataType, OutputDataType>&
+Reparametrization<InputDataType, OutputDataType>::
+opertaor=(Reparametrization&& layer)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/reparametrization_impl.hpp:90]:
```suggestion
operator=(Reparametrization&& layer) 
```

Minor spelling fix.

File: src/mlpack/methods/ann/layer/reparametrization_impl.hpp [hunk 2]
```diff
@@ -46,7 +46,60 @@ Reparametrization<InputDataType, OutputDataType>::Reparametrization(
...
+    includeKl(std::move(layer.includeKl)),
+    beta(std::move(layer.beta))
+{
+   //Nothing to do here     
+}
+    
+template <typename InputDataType, typename OutputDataType>
+Reparametrization<InputDataType, OutputDataType>&
+Reparametrization<InputDataType, OutputDataType>::
+opertaor=(const Reparametrization& layer)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/reparametrization_impl.hpp:75]:
```suggestion
operator=(const Reparametrization& layer) 
```

Minor spelling fix.

File: src/mlpack/methods/ann/layer/reparametrization.hpp [hunk 3]
```diff
@@ -71,6 +71,18 @@ class Reparametrization
...
                     const bool includeKl = true,
                     const double beta = 1);
+    
+  //! Copy Constructor 
+  Reparametrization(const Reparametrization& layer);
+    
+  //! Move Constructor 
+  Reparametrization(Reparametrization&& layer);
+    
+  //! Copy assignment operator
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/reparametrization.hpp:81]:
```suggestion
  //! Copy Constructor.
  Reparametrization(const Reparametrization& layer);
    
  //! Move Constructor.
  Reparametrization(Reparametrization&& layer);
    
  //! Copy assignment operator.
```

Add missing stop at the end.

**Comment by zoq** [comment]
To test this locally it would be a good idea to build with `cmake -DDEBUG=ON` to enable bounds checks.

**Comment by Anmol2001** [comment]
> To test this locally it would be a good idea to build with `cmake -DDEBUG=ON` to enable bounds checks.

I have updated the test cases and hopefull test cases will pass

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 4]
```diff
@@ -153,6 +153,40 @@ TEST_CASE("CheckCopyMovingVanillaNetworkTest", "[FeedForwardNetworkTest]")
...
+  trainData.shed_row(trainData.n_rows - 1);
+
+  /*
+   * Construct a feed forward network with trainData.n_rows input nodes,
+   * followed by a linear layer and then a reparametrization layer
+   */
+
+  FFN<NegativeLogLikelihood<> > *model = new FFN<NegativeLogLikelihood<> >;
+  model->Add<Linear<> >(trainData.n_rows, 8);
+  model->Add<Reparametrization<> >(4,false,true,1);
```
Comment (zoq) [inline src/mlpack/tests/feedforward_network_test.cpp:175]:
```suggestion
  model->Add<Reparametrization<> >(4, false, true, 1);
```

Minor style issue, missing spaces.

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 5]
```diff
@@ -153,6 +153,40 @@ TEST_CASE("CheckCopyMovingVanillaNetworkTest", "[FeedForwardNetworkTest]")
...
+  model->Add<LogSoftMax<> >();
+
+  FFN<NegativeLogLikelihood<> > *model1 = new FFN<NegativeLogLikelihood<> >;
+  model1->Add<Linear<> >(trainData.n_rows, 8);
+  model1->Add<Reparametrization<> >(4,false,true,1);
```
Comment (zoq) [inline src/mlpack/tests/feedforward_network_test.cpp:180]:
```suggestion
  model1->Add<Reparametrization<> >(4, false, true, 1);
```

Minor style issue, missing spaces.

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 6]
```diff
@@ -153,6 +153,40 @@ TEST_CASE("CheckCopyMovingVanillaNetworkTest", "[FeedForwardNetworkTest]")
...
+  // Load the dataset.
+  arma::mat trainData;
+  data::Load("thyroid_train.csv", trainData, true);
+
+  arma::mat trainLabels = trainData.row(trainData.n_rows - 1);
+  trainData.shed_row(trainData.n_rows - 1);
+
+  /*
+   * Construct a feed forward network with trainData.n_rows input nodes,
+   * followed by a linear layer and then a reparametrization layer
```
Comment (zoq) [inline src/mlpack/tests/feedforward_network_test.cpp:170]:
```suggestion
   * followed by a linear layer and then a reparametrization layer.
```

Missing stop at the end.

File: src/mlpack/methods/ann/layer/reparametrization_impl.hpp [hunk 7]
```diff
@@ -46,7 +46,60 @@ Reparametrization<InputDataType, OutputDataType>::Reparametrization(
...
+operator=(Reparametrization&& layer) 
+{
+       if (this != &layer)
+       {
+         latentSize = std::move(layer.latentSize);
+         stochastic = std::move(layer.stochastic);
+         includeKl = std::move(layer.includeKl);
+         beta = std::move(layer.beta);
+       }
+       return *this;
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/reparametrization_impl.hpp:99]: The indentation is off, should be one space.

Comment (Anmol2001) [reply src/mlpack/methods/ann/layer/reparametrization_impl.hpp]: Thanks for review.. done

File: src/mlpack/methods/ann/layer/reparametrization.hpp [hunk 8]
```diff
@@ -71,6 +71,18 @@ class Reparametrization
...
+  //! Copy Constructor.
+  Reparametrization(const Reparametrization& layer);
+    
+  //! Move Constructor.
+  Reparametrization(Reparametrization&& layer);
+    
+  //! Copy assignment operator.
+  Reparametrization& operator=(const Reparametrization& layer);
+    
+  //! Move assignment operator
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/reparametrization.hpp:84]:
```suggestion
  //! Move assignment operator.
```

One more minor style issue, missing stop at the end.

File: src/mlpack/methods/ann/layer/reparametrization_impl.hpp [hunk 9]
```diff
@@ -46,7 +46,60 @@ Reparametrization<InputDataType, OutputDataType>::Reparametrization(
...
+    
+template <typename InputDataType, typename OutputDataType>
+Reparametrization<InputDataType, OutputDataType>::Reparametrization(
+    const Reparametrization& layer) :
+    latentSize(layer.latentSize),
+    stochastic(layer.stochastic),
+    includeKl(layer.includeKl),
+    beta(layer.beta)
+{
+   //Nothing to do here
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/reparametrization_impl.hpp:58]:
```suggestion
   // Nothing to do here.
```

Pedantic style issue, missing stop and remove trailing spaces.

File: src/mlpack/methods/ann/layer/reparametrization_impl.hpp [hunk 10]
```diff
@@ -46,7 +46,60 @@ Reparametrization<InputDataType, OutputDataType>::Reparametrization(
...
 
+template <typename InputDataType, typename OutputDataType>
+Reparametrization<InputDataType, OutputDataType>::Reparametrization(
+    Reparametrization&& layer) :
+    latentSize(std::move(layer.latentSize)),
+    stochastic(std::move(layer.stochastic)),
+    includeKl(std::move(layer.includeKl)),
+    beta(std::move(layer.beta))
+{
+   //Nothing to do here
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/reparametrization_impl.hpp:69]:
```suggestion
   // Nothing to do here.
```

Pedantic style issue, missing stop and remove trailing spaces.

**Review by zoq — COMMENTED** [review]
Thanks was missing some last minor style issues.

**Comment by Anmol2001** [comment]
@zoq  i have done the suggested changes : )

**Review by zoq — APPROVED** [review]
Thanks, no more comments from my side.

**Comment by zoq** [comment]
Thanks for the contribution.

**Pull request merged**